source/metrics/write_to_csv.cpp

```diff
@@ -2,54 +2,103 @@
 
 #include <cmath>
 #include <limits>
+#include <span>
+
+static constexpr std::size_t FLUSH_THRESHOLD = (1 << 20);  // 1 MB
 
 namespace sim {
 
 void write_to_csv(
-    const std::vector<std::pair<MetricsStorage, std::string> >& storages,
-    std::filesystem::path output_path) {
-    size_t count_storages = storages.size();
-    // values[time][i] is a value of metric for i-th storage at time time;
-    // If there were no measurement at time, values[time][i] =
-    // std::numeric_limits<double>::quiet_NaN()
-    std::map<TimeNs, std::vector<double> > values;
-    double nan = std::numeric_limits<double>::quiet_NaN();
-    std::vector<double> default_values(count_storages, nan);
-    for (size_t i = 0; i < count_storages; i++) {
-        for (const auto& [time, value] : storages[i].first.get_records()) {
-            if (values.find(time) == values.end()) {
-                values[time] = default_values;
+    const std::vector<std::pair<MetricsStorage, std::string>>& storages,
+    const std::filesystem::path output_path) {
+    const size_t count_storages = storages.size();
+
+    using Sample = std::pair<TimeNs, double>;
+    using SeriesSpan = std::span<const Sample>;
+
+    // References to the underlying data vectors
+    std::vector<SeriesSpan> series;
+    series.reserve(count_storages);
+
+    for (const auto& [storage, name] : storages) {
+        const std::vector<Sample>& recs = storage.get_records();
+        series.emplace_back(recs.data(), recs.size());
+    }
+
+    // Indices of the current element for each storage
+    std::vector<size_t> idx(count_storages, 0);
+
+    const double nan = std::numeric_limits<double>::quiet_NaN();
+    // Most recent known values (forward-fill state)
+    std::vector<double> values(count_storages, nan);
+
+    utils::create_all_directories(output_path);
+    std::ofstream out(output_path, std::ios::binary);
+    if (!out) {
+        throw std::runtime_error("Failed to open output file: " +
+                                 output_path.string());
+    }
+
+    fmt::memory_buffer buffer;
+    auto it = std::back_inserter(buffer);
+
+    // Header row
+    fmt::format_to(it, "Time");
+    for (const auto& p : storages) {
+        fmt::format_to(it, ",{}", p.second);
+    }
+    fmt::format_to(it, "\n");
+
+    auto get_next_time = [&]() -> std::optional<TimeNs> {
+        std::optional<TimeNs> result;
+        for (size_t i = 0; i < count_storages; ++i) {
+            const auto& recs = series[i];
+            if (idx[i] < recs.size()) {
+                TimeNs t = recs[idx[i]].first;
+                if (!result || t < result.value()) {
+                    result = t;
+                }
             }
-            values[time][i] = value;
         }
-    }
+        return result;
+    };
+
+    auto flush_buffer = [&]() {
+        out.write(buffer.data(), static_cast<std::streamsize>(buffer.size()));
```

Comment:
Clear buffer in flush
```suggestion
        out.write(buffer.data(), static_cast<std::streamsize>(buffer.size()));
        buffer.clear();
```